### ISSUE TITLE
gather算子在最后一个轴的性能与AscendC算子性能差距过大

### ISSUE THREAD

**Issue by chaosisnotopen** [triton-lang/triton-ascend#339]
# Copyright © 2026 Huawei Technologies Co., Ltd.
import math
import torch
import triton
import triton.language as tl
import numpy as np

DEVICE = (
    "cuda"
    if torch.cuda.is_available()
    else "npu"
    if torch.npu.is_available()
    else "cpu"
)

@triton.jit
def _triton_gather_notune(
    y_ptr,
    x_ptr,
    indices_ptr,
    y_outer_size: tl.constexpr,
    x_axis_size: tl.constexpr,
    inner_size: tl.constexpr,
    indices_size: tl.constexpr,
    x_stride0: tl.constexpr,
    x_stride1: tl.constexpr,
    y_stride0: tl.constexpr,
    y_stride1: tl.constexpr,
    BLOCK_N: tl.constexpr,
    BLOCK_SIZE: tl.constexpr,
):
    pid = tl.program_id(0).to(tl.int32)
    num_core = tl.num_programs(0)
    num_blocks = tl.cdiv(inner_size, BLOCK_SIZE)

    for iN in range(BLOCK_N):
        gid = pid + iN * num_core

        for h in range(indices_size):
            index = tl.load(indices_ptr + h)

            for block_idx in range(num_blocks):
                offset = block_idx * BLOCK_SIZE
                mask = (offset + tl.arange(0, BLOCK_SIZE) < inner_size) & (
                    gid < y_outer_size
                )

                x_off = gid * x_stride0 + index * x_stride1 + offset
                x = tl.load(
                    x_ptr + x_off + tl.arange(0, BLOCK_SIZE), mask=mask, other=0.0
                )

                y_off = gid * y_stride0 + h * y_stride1 + offset
                tl.store(y_ptr + y_off + tl.arange(0, BLOCK_SIZE), x, mask=mask)

@triton.jit
def _triton_gather_notune_axisn(
    y_ptr,
    x_ptr,
    indices_ptr,
    y_outer_size: tl.constexpr,
    x_axis_size: tl.constexpr,
    inner_size: tl.constexpr,
    indices_size: tl.constexpr,
    x_stride0: tl.constexpr,
    x_stride1: tl.constexpr,
    y_stride0: tl.constexpr,
    y_stride1: tl.constexpr,
    BLOCK_N: tl.constexpr,
    BLOCK_SIZE: tl.constexpr,
):
    pid = tl.program_id(0).to(tl.int32)
    num_programs = tl.num_programs(0)

    for iN in range(BLOCK_N):
        gid = pid + iN * num_programs
        mask = gid < y_outer_size

        indices = tl.load(indices_ptr + tl.arange(0, indices_size), mask=mask)

        base_x = gid * x_stride0
        val = tl.load(x_ptr + base_x + indices, mask=mask)

        base_y = gid * y_stride0
        tl.store(y_ptr + base_y + tl.arange(0, indices_size), val, mask=mask)

@triton.autotune(
    configs=[
        triton.Config({"BLOCK_SIZE": 128}),
        triton.Config({"BLOCK_SIZE": 256}),
        triton.Config({"BLOCK_SIZE": 512}),
        triton.Config({"BLOCK_SIZE": 1024}),
        triton.Config({"BLOCK_SIZE": 2048}),
        triton.Config({"BLOCK_SIZE": 4096}),
    ],
    key=["inner_size"],
)
@triton.jit
def _triton_gather_autotune(
    y_ptr,
    x_ptr,
    indices_ptr,
    y_outer_size: tl.constexpr,
    x_axis_size: tl.constexpr,
    inner_size: tl.constexpr,
    indices_size: tl.constexpr,
    x_stride0: tl.constexpr,
    x_stride1: tl.constexpr,
    y_stride0: tl.constexpr,
    y_stride1: tl.constexpr,
    BLOCK_N: tl.constexpr,
    BLOCK_SIZE: tl.constexpr,
):
    pid = tl.program_id(0).to(tl.int32)
    num_core = tl.num_programs(0)
    num_blocks = tl.cdiv(inner_size, BLOCK_SIZE)

    for iN in range(BLOCK_N):
        gid = pid + iN * num_core

        for h in range(indices_size):
            index = tl.load(indices_ptr + h)

            for block_idx in range(num_blocks):
                offset = block_idx * BLOCK_SIZE
                mask = (offset + tl.arange(0, BLOCK_SIZE) < inner_size) & (
                    gid < y_outer_size
                )

                x_off = gid * x_stride0 + index * x_stride1 + offset
                x = tl.load(
                    x_ptr + x_off + tl.arange(0, BLOCK_SIZE), mask=mask, other=0.0
                )

                y_off = gid * y_stride0 + h * y_stride1 + offset
                tl.store(y_ptr + y_off + tl.arange(0, BLOCK_SIZE), x, mask=mask)

def triton_gather(
    x, indices, axis, batch_dims=0, negative_index_support=False, block_size=0
):
    x = x.to(DEVICE)
    indices = indices.to(DEVICE)
    x_outer_size = math.prod(x.shape[0:axis])
    x_inner_size = math.prod(x.shape[axis + 1 :])
    x_axis_size = x.shape[axis]

    out_shape = list(x.shape)
    out_shape[axis] = len(indices)
    y_outer_size = math.prod(out_shape[0:axis])
    indices_size = len(indices)
    inner_size = x_inner_size

    y = torch.empty(out_shape, device=x.device, dtype=x.dtype)

    x_strides = (x_axis_size * inner_size, inner_size, 1)
    y_strides = (indices_size * inner_size, inner_size, 1)

    if DEVICE == "cuda":
        grid = (y_outer_size,)
        BLOCK_N = 1
        block_size = triton.next_power_of_2(block_size)  # support cuda
    else:
        import triton.runtime.driver as driver

        core_num = driver.active.utils.get_device_properties(
            torch.npu.current_device()
        )["num_vectorcore"]
        grid = (min(y_outer_size, core_num),)
        BLOCK_N = (y_outer_size + grid[0] - 1) // grid[0]

    if block_size == 0:
        _triton_gather_autotune[grid](
            y,
            x,
            indices,
            y_outer_size,
            x_axis_size,
            inner_size,
            indices_size,
            x_strides[0],
            x_strides[1],
            y_strides[0],
            y_strides[1],
            BLOCK_N,
        )
    else:
        if axis == len(x.shape) - 1:
            block_size = inner_size
            _triton_gather_notune_axisn[grid](
                y,
                x,
                indices,
                y_outer_size,
                x_axis_size,
                inner_size,
                indices_size,
                x_strides[0],
                x_strides[1],
                y_strides[0],
                y_strides[1],
                BLOCK_N,
                block_size,
            )
        else:
            # ub overflow
            max_size = 65536 // torch.empty((), dtype=x.dtype).element_size()
            if block_size > max_size:
                if inner_size < max_size:
                    block_size = inner_size
                else:
                    block_size = max_size
            _triton_gather_notune[grid](
                y,
                x,
                indices,
                y_outer_size,
                x_axis_size,
                inner_size,
                indices_size,
                x_strides[0],
                x_strides[1],
                y_strides[0],
                y_strides[1],
                BLOCK_N,
                block_size,
            )
    return y

测试代码如下：

import sys
from itertools import product
from pathlib import Path
import torch
import triton
import pytest

ROOT = Path(__file__).parent.parent
sys.path.insert(0, str(ROOT))

from kernel.gather import triton_gather

DEVICE = (
    "cuda"
    if torch.cuda.is_available()
    else "npu"
    if torch.npu.is_available()
    else "cpu"
)
SHAPE_SIZE = [1, 2, 3, 4]
SHAPE_DIM_MIN = [2]
SHAPE_DIM_MAX = [150]
DTYPES = [torch.float32, torch.float16, torch.int8]
SEEDS = [123]
PERF_SHAPES = [[8, 16, 32, 5915], [16, 8, 600, 32], [64, 8, 16, 32]]
# PERF_SHAPES = [[8, 128, 2050, 9155], [8, 64, 1120, 300], [64, 8, 128, 32]]
PERF_AXIS = [3, 2, 0]

# 精度标准
RTOL = {torch.float32: 1e-3, torch.float16: 1e-1}
ATOL = {torch.float32: 1e-4, torch.float16: 1e-1}

# 性能测试标准
PERFORMANCE_THRESHOLD = 0.9  # 0.9 * AscendC
WARMUP = 0
ITER = 1

def torch_gather(params: torch.Tensor, indices: torch.Tensor, axis=0, batch_dims=0):
    out = torch.index_select(params, axis, indices)
    return out

def gen_rand_tensor(shape, dtype):
    if dtype in [torch.float16, torch.float32, torch.bfloat16]:
        params = (
            torch.empty(shape, dtype=dtype).uniform_(-1.0, 1.0) * torch.finfo(dtype).max
        )
    elif dtype in [
        torch.int32,
        torch.uint32,
        torch.int16,
        torch.uint16,
        torch.int8,
        torch.uint8,
    ]:
        info = torch.iinfo(dtype)
        params = torch.randint(info.min, info.max, shape, dtype=dtype)
    else:
        raise TypeError("Invalid dtype {}".format(dtype))

    return params

def benchmark_perf(shape, indices_size, axis, dtype, block_size, provider):
    torch.manual_seed(123)
    x = gen_rand_tensor(shape, dtype).to(DEVICE)
    indices = torch.randint(0, shape[axis], (indices_size,), dtype=torch.int32).to(
        DEVICE
    )

    USE_AUTOTUNE = block_size == 0

    if provider == "torch":
        ms = triton.testing.do_bench(lambda: torch_gather(x, indices, axis))
    if provider == "triton":
        if USE_AUTOTUNE:
            ms = triton.testing.do_bench(lambda: triton_gather(x, indices, axis))
        else:
            ms = triton.testing.do_bench(
                lambda: triton_gather(x, indices, axis, block_size=block_size)
            )

    return ms

def run_perf_testcase(caseId, perfMod, block_size, dtype_id):
    """性能测试函数"""
    """caseId: 0: large, 1: middle, 2: small"""
    """perfMode: bit0: run triton, bit1: run torch """
    """block_size = 0: autotune """
    assert (
        caseId == 0 or caseId == 1 or caseId == 2
    ), "Error input caseId={}, expect in [0, 1, 2]".format(caseId)

    # 性能测试使用最大矩阵，最能体现性能
    shape = PERF_SHAPES[caseId]  # caseId: 0: large, 1: middle, 2: small
    axis = PERF_AXIS[caseId]
    indices_size = shape[axis] // 3
    dtype = DTYPES[dtype_id]

    print("\n" + "=" * 60)
    print(
        " Triton gather 性能测试 caseId={} BLOCK_SIZE={} dtype={}".format(
            caseId, block_size, dtype
        )
    )
    print("=" * 60)

    """perfMode: bit0: run triton, bit1: run torch"""
    # 测速
    t_torch = (
        benchmark_perf(shape, indices_size, axis, dtype, block_size, "torch")
        if (perfMod % 2) == 0
        else 0
    )

    t_triton = (
        benchmark_perf(shape, indices_size, axis, dtype, block_size, "triton")
        if perfMod % 2
        else 1e-12
    )

    # 性能倍率
    ratio = t_torch / t_triton
    meet_standard = ratio >= PERFORMANCE_THRESHOLD

    # 输出
    print(f"[矩阵尺寸] shape={shape}, axis={axis}, BLOCK_SIZE={block_size}, DTYPE={dtype}")
    print(f"[数据类型] {dtype}")
    print(f"[PyTorch]  平均耗时: {t_torch:>.3f} ms")
    print(f"[Triton]   平均耗时: {t_triton:>.3f} ms")
    print(f"[性能倍率] {ratio:.3f}x")
    print(f"[达标 0.9×AscendC] : {'✅ 满足' if meet_standard else '❌ 不满足'}")
    print("=" * 60 + "\n")

if __name__ == "__main__":
        caseId = int(sys.argv[1]) if len(sys.argv) >= 2 else 0
        perfMod = int(sys.argv[2]) if len(sys.argv) >= 3 else 3
        block_size = int(sys.argv[3]) if len(sys.argv) >= 4 else 128
        dtype_id = int(sys.argv[4]) if len(sys.argv) >= 5 else 0
        run_perf_testcase(caseId, perfMod, block_size, dtype_id)

950PR上的测试结果：
msprof的结果如下：
                     AscendC          Triton
总耗时(ms)   0.0995              9.3855

</body>

</html>

### Environment details

Triton 3.2.1

cann9.0.0
950PR

**Comment by WuTYSFG** [comment]
Thanks for your feedback. We will arrange relevant engineers to investigate this issue later.
